File: OCT13/SEATRSF.cpp
#include <iostream>
using namespace std;

#define ll long long int
#define M 1000000007

int main(){
  ll t,n,m,q,k,i,j;
  //  cout << 1000000000LL*1000000000LL << endl;
  cin >> t;
  while(t--){
    cin >> n >> m >> q >> k;
    i = 1;
    ll ans1 = -1, ans = 0;
    while(q+i <= m){
      ans1 = (n*(n-1)/2)%M;
      for(j=0;j<n-2;j++)
	ans1 = ( (ans1%M)*((q+1)%M) )%M;
      ans = ( (ans%M) + (ans1)%M )%M;
      i++;
    }
    if(n == 1 || ans1<0){
      cout << 0 << endl;
    }
    else{
      cout << ans << endl;
    }
  }
  return 0;
}

File: OCT13/SEATRSF.py
M = 1000000007;
def power(a, b):
    if b == 0:
        return 1;
    elif b == 1:
        return a;
    else:
        t = ( power(a,b/2) )%M;
        if b%2 == 1:
            return (((t*t)%M)*a)%M;
        else:
            return (t*t)%M;
t = int(raw_input());
while t != 0:
    n,m,q,k = map(int,raw_input().split(" "));
    if n == 1 or  m <= q:
        print 0;
    else:
        ans1 = ( (power(q+1,n)%M) - (2*power(q,n))%M + (power(q-1,n))%M )%M;
        ans = ( (m-q)*(ans1) )%M;
        print ans;
    t -= 1;
